Feature/makefile

mjkrause opened the pull request:
* makefile with a deploy and a teardown function
* * deploy: creates a new direction in the present directory with a copy of the repo but the new name of the Lambda function
* * teardown: executes the chalice delete method (AWS Lambda function, API Gateway, IAM role), deactivates the virtual environment, and deletes the directory

┆Issue Number: BFL-4

mjkrause committed:
* makefile first pass
* added a makefile
* added paragraph `Deployment` in `README`

mjkrause committed:
corrected makefile and README

mjkrause committed:
added line to makefile to copy config to new chalice folder

mjkrause committed:
corrected last line added to makefile

mjkrause committed:
correction in README

mjkrause committed:
* added line to change lambda name (using `jq`)
* corrected README

mjkrause committed:
* corrected `config.json`

mjkrause committed:
* added shell script to edit lambda name in config.json

mjkrause committed:
* corrected config.json

mjkrause committed:
* removed unnecessary line in teardown function of makefile

mjkrause committed:
* changed makefile to exec. shell script

mjkrause committed:
* changed makefile to exec. shell script

mjkrause committed:
* changed makefile to exec. shell script

mjkrause commented:
A few points with respect to the design I've pondering over and where some opinion / input would be helpful:
1. current implementation uses the `jq` utility, which is installed by default on Ubuntu but potentially not on other OS's (OSX?)
2. I'm wondering whether the makefile should instead of creating a new directory inside the repo (using the `chalice new ...` command) just rename the actual repo directory (which is `bagit-firecloud-lambda`) into the name chosen by the user. In that case we would just edit the chalice configuration file.

mjkrause committed:
* added deactivation of virtual environment to target teardown

mjkrause committed:
* corrected tab

mjkrause committed:
* corrected deactivate command

mjkrause committed:
* removed deactivating virtual environment from teardown target

mjkrause committed:
* adding tests to make file

mjkrause committed:
* completely new Makefile
   * user only needs to clone the repo and run `make deploy`
   * it creates a virtual environment
   * creates a dev and a staging stage
   * runs all tests
* added `common.mk` which contains a general check whether Python 3.6
is installed

mjkrause committed:
* (forgot to add `common.mk`)

mjkrause committed:
* changed comments in Makefile

coverbeck commented on common.mk:
```diff
@@ -0,0 +1,5 @@
+SHELL=/bin/bash
+
+ifeq ($(findstring python3.6, $(shell ls -ls /usr/bin/python* | grep python3.6 2>&1)),)
```
Do you need the trailing comma (next to last character)?

mjkrause replied:
Yes, that comma is needed as it seems part of the GNU `make` syntax. Throws an error without it.

coverbeck replied:
Got it.

I was trying this, and I don't think your test for Python 3.6 is correct, at least not on my Mac. There is a space between `Python` and `3.6`, and my Python has a capital P.

```
(.virtualenvs2) ~/git/DataBiosphere/bagit-firecloud-lambda$ make
common.mk:4: *** Please install Python 3.6 on the system.  Stop.
(.virtualenvs2) ~/git/DataBiosphere/bagit-firecloud-lambda$ python
Python 3.6.4 (default, Mar  1 2018, 18:36:42) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```

mjkrause replied:
could you please post the output of `ls -ls /usr/bin/python*` from your Mac?

coverbeck replied:
Not sure if this is a typical set up or not... Also added `which python`
```
~$ ls -ls /usr/bin/python*
32 -rwxr-xr-x  1 root  wheel  66576 Jul 14  2017 /usr/bin/python
 0 -rwxr-xr-x  5 root  wheel    925 Feb  7  2017 /usr/bin/python-config
 8 lrwxr-xr-x  1 root  wheel     75 Jul 15  2017 /usr/bin/python2.6 -> ../../System/Library/Frameworks/Python.framework/Versions/2.6/bin/python2.6
 8 lrwxr-xr-x  1 root  wheel     82 Jul 15  2017 /usr/bin/python2.6-config -> ../../System/Library/Frameworks/Python.framework/Versions/2.6/bin/python2.6-config
 8 lrwxr-xr-x  1 root  wheel     75 Jul 15  2017 /usr/bin/python2.7 -> ../../System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7
 8 lrwxr-xr-x  1 root  wheel     82 Jul 15  2017 /usr/bin/python2.7-config -> ../../System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7-config
32 -rwxr-xr-x  1 root  wheel  66576 Jul 14  2017 /usr/bin/pythonw
 8 lrwxr-xr-x  1 root  wheel     76 Jul 15  2017 /usr/bin/pythonw2.6 -> ../../System/Library/Frameworks/Python.framework/Versions/2.6/bin/pythonw2.6
 8 lrwxr-xr-x  1 root  wheel     76 Jul 15  2017 /usr/bin/pythonw2.7 -> ../../System/Library/Frameworks/Python.framework/Versions/2.7/bin/pythonw2.7
~$ which python
/anaconda3/bin/python
```

coverbeck commented on common.mk:
```diff
@@ -0,0 +1,5 @@
+SHELL=/bin/bash
+
+ifeq ($(findstring python3.6, $(shell ls -ls /usr/bin/python* | grep python3.6 2>&1)),)
+$(error Please install Python 3.6 on the system)
```
What is `error`? I don't have that on my Mac. Is it a Linux command?

I would think you would to exit with a non-zero exit code here, although maybe error already does that.

mjkrause replied:
`error` is a make keyword. If the equivalence is false it return to stdout with that error message. Do you have GNU Make on your Mac? If so that command should exist.

coverbeck replied:
Got it; my bad.

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+	.venvbag/bin/pip install awscli --upgrade
+	@echo "deploying the Lambda"
+	.venvbag/bin/chalice deploy --stage dev
+	.venvbag/bin/chalice deploy --stage staging
```
I think you should have separate deploys for dev and staging. The typical flow would be to deploy to dev repeatedly until you are sure it is working correctly, then deploy to staging when you are confident it works.

mjkrause replied:
do you mean like a retry on the `chalice deploy`? perhaps it does it already internally? I can research it. I can separate deployment of the two stages.

coverbeck replied:
I mean you should have separate targets to deploy dev and to deploy stage, so that the invoker of make can choose which env to deploy to.

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+venv/bin/activate: requirements.txt
+	test -d .venvbag || virtualenv --python=python3.6 .venvbag
+	.venvbag/bin/pip install -Ur requirements.txt
+	touch .venvbag/bin/activate
```
What is the purpose of this?

mjkrause replied:
deployment of the lambda depends on python packages which are not standard. usually this is done by creating a virtual environment and installing those packages (which are listed in `requirements.txt`). This construct creates such a virtual environment, and installs requirements and deploys the lambda inside of it. This way user action is minimized.

coverbeck replied:
Sorry, I was not clear -- what is the purpose of the `touch` line? I understand the virtualenv and requirements part.

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+	touch .venvbag/bin/activate
+
+deploy: venv
+	@$(MAKE) -f $(THIS_FILE)  all
```
Could you just make `all` a dependency?

mjkrause replied:
the target `deploy` depends on both, `venv` and `all`. But I understand you can only create at dependency on a single target without using the `$(MAKE)` construct (i.e., the way I use it for `venv`). but because there are two dependencies I had use `$(MAKE)` for the other.

coverbeck replied:
If I have:

```
deploy: venv all
```
And get rid of the `@$(MAKE)...` line, all gets executed.

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+
+ALLMODULES=$(patsubst %.py, %, $(wildcard test_*.py))
+
+all:
```
Why not call this something like `tests` instead of `all`?

mjkrause replied:
pl. see my response below

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+	@echo "running tests..."
+	${RUNTEST} ${ALLMODULES}
+
+% : test_%.py
```
Does this do anything? I don't see it being invoked, although I may be missing something.

mjkrause replied:
I found this with `all` on the web, renamed it to either `test` or `tests`, and it stopped working, so I put `all` back in. The target `all` is a convention (see https://www.gnu.org/software/make/manual/html_node/Standard-Targets.html#Standard-Targets). However, in this Makefile I wouldn't consider `all` to compile the entire program (doesn't compile anything actually), so it isn't appropriate. I find out which one of the two test/tests works and try to replace it.

coverbeck replied:
If all is the convention, I don't think it should be on the tests.

I think all should do everything short of the deploy -- install the virtualenv and run the tests.

If someone clones the repo from scratch and just does a `make`, won't it fail? Because that will just run the tests, but will have not set up the virtualenv...

I agree on the need for `all`, especially after reading the link you provided, but I don't think it's functionality should be to run the tests.

coverbeck commented on README.md:
```diff
@@ -6,8 +6,15 @@ A staging instance is currently stood up at `https://vzltaytzg4.execute-api.us-w
 
+2. Run `make deploy`. 
+
+This will install two stages of the lambda, `-dev` and one `-staging`. It will deploy a `dev` and a `staging` stage (two separate Lambdas). After you run `make deploy` a subdirectory `.venvbag` has been created in that directory. If you want to delete any of them run for instance `.venvbag/bin/chalice delete --stage dev` to delete the `dev` stage. 
```
Should there be a target for this instead? I'm not sure on this one.

mjkrause replied:
not sure what you mean by "for this...". target for what?

mjkrause replied:
Ah, you mean the delete function? I tried to a teardown that deletes both targets, but it error-ed on deleting the second lambda. perhaps it would work reliable if we separate them into individual targets. It wouldn't make sense in terms of saving number of characters to type: compare `.venvbag/bin/chalice delete --stage dev` to `make teardown-dev`.

coverbeck reviewed (changes requested):
Several of my comments are questions, as I'm not sure if I'm missing something.

mjkrause committed:
* changes to Makefile: runs all tests, verified that it aborts
deployment if any test fails, and created separate targets for dev and
 staging (removed target all)

* small changes in `test_manifest.py`

mjkrause committed:
* changes to Makefile: runs all tests, verified that it aborts
deployment if any test fails, and created separate targets for dev and
 staging (removed target all)

mjkrause committed:
* changes to Makefile: created new target `tests` and made it a
dependency in both deploy targets.

mjkrause committed:
* updated README

coverbeck commented on Makefile:
```diff
@@ -14,19 +13,15 @@ venv/bin/activate: requirements.txt
+tests: venv
+	@echo "in tests"
 	.venvbag/bin/python ${RUNTEST} ${ALLMODULES}
 	.venvbag/bin/pip install awscli --upgrade
```
While I understand wanting to avoid duplication, it is still confusing to have a target called tests that installs the AWS client.

Can you create another target that has a dependency on tests? Something like:

```
tests: venv
     @echo "Running tests"
      ....

aws-cli: tests
   .venvbag/bin/pip install awscli --upgrade

deploy-dev: aws-cli

coverbeck commented on Makefile:
```diff
@@ -14,19 +13,15 @@ venv/bin/activate: requirements.txt
-deploy-dev: venv
-	@echo "running tests..."
+tests: venv
+	@echo "in tests"
```
Should say running tests

mjkrause committed:
* changed from `httpbin.org` to `httpstat.us` to get status code for
test (response is much faster)
* created new target in Makefile `awssetup` that upgrades awscli
* moved all test data in a new subdirectory
* split `utils.py` into `utils.py` and `manifest_io.py` so they
correspond to the respective test files
   * corrected all paths and import statements
* all tests pass
* tested Makefile

coverbeck commented on Makefile:
```diff
@@ -0,0 +1,31 @@
+	touch .venvbag/bin/activate
+
+tests: venv
+	@echo "in tests"
```
I would say `Running tests`

mjkrause replied:
done

coverbeck reviewed (approved):
One small suggestion on wording, but looks good.

mjkrause committed:
* changed comment to stdout in Makefile
